src/table.mjs

```diff
@@ -1,7 +1,27 @@
+import {ascending, descending, reverse} from "d3-array";
+
+export function isTypedArray(value) {
+  return (
+    value instanceof Int8Array ||
+    value instanceof Int16Array ||
+    value instanceof Int32Array ||
+    value instanceof Uint8Array ||
+    value instanceof Uint8ClampedArray ||
+    value instanceof Uint16Array ||
+    value instanceof Uint32Array ||
+    value instanceof Float32Array
+  );
+}
+
 export const __query = Object.assign(
   // This function is used by table cells.
   async (source, operations, invalidation) => {
-    const args = makeQueryTemplate(operations, await source);
+    // For cells whose data source is an in-memory table, we use JavaScript to
+    // apply the table cell operations, instead of composing a SQL query.
+    // TODO Do we need to port over all the logic from canDisplayTable, or is
```

Comment:
[canDisplayTable](https://github.com/observablehq/observablehq/blob/main/notebook-next/src/worker/display.js#L24) contains a lot of logic, and i don't know if we need to copy that over to stdlib? it seems like checking the general structure of `source` is sufficient here, since the source needs to have passed the `canDisplayTable` check to even show up as a data source option in the first place.

Reply:
I’ve ported this logic over so that we can show the correct errors and also handle promotion of primitive arrays.